sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala

```diff
@@ -0,0 +1,185 @@
+      val filters = splitConjunctivePredicates(deleteExpr).flatMap { p =>
+        val filter = DataSourceStrategy.translateFilter(p, supportNestedPredicatePushdown = true)
+        if (filter.isEmpty) {
+          throw new AnalysisException(s"Cannot translate expression to source filter: $p")
+        }
+        filter
+      }.toArray
+
+      val table = r.table
+      val writeBuilder = newWriteBuilder(table, query, options)
+      val write = writeBuilder match {
+        case builder: SupportsTruncate if isTruncate(filters) =>
+          builder.truncate().build()
+        case builder: SupportsOverwrite =>
+          builder.overwrite(filters).build()
+        case _ =>
+          throw new SparkException(s"Table does not support overwrite by expression: $table")
+      }
+
+      o.copy(write = Some(write), query = addDistributionAndOrdering(write, query))
+
+    case o @ OverwritePartitionsDynamic(r: DataSourceV2Relation, query, options, _, None) =>
+      val table = r.table
+      val writeBuilder = newWriteBuilder(table, query, options)
+      val write = writeBuilder match {
+        case builder: SupportsDynamicOverwrite =>
+          builder.overwriteDynamicPartitions().build()
+        case _ =>
+          throw new SparkException(s"Table does not support dynamic partition overwrite: $table")
+      }
+      o.copy(write = Some(write), query = addDistributionAndOrdering(write, query))
+  }
+
+  private def newWriteBuilder(
+      table: Table,
+      query: LogicalPlan,
+      writeOptions: Map[String, String]): WriteBuilder = {
+
+    val info = LogicalWriteInfoImpl(
+      queryId = UUID.randomUUID().toString,
+      query.schema,
+      writeOptions.asOptions)
+    table.asWritable.newWriteBuilder(info)
+  }
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
+  }
+
+  private def addDistributionAndOrdering(
+      write: Write,
+      query: LogicalPlan): LogicalPlan = write match {
+
+    case write: RequiresDistributionAndOrdering =>
+      val sqlConf = SQLConf.get
+      val resolver = sqlConf.resolver
+
+      val distribution = write.requiredDistribution match {
+        case d: OrderedDistribution =>
+          d.ordering.map(e => toCatalyst(e, query, resolver))
+        case d: ClusteredDistribution =>
+          d.clustering.map(e => toCatalyst(e, query, resolver))
+        case _: UnspecifiedDistribution =>
+          Array.empty[catalyst.expressions.Expression]
+      }
+
+      val queryWithDistribution = if (distribution.nonEmpty) {
+        val numShufflePartitions = sqlConf.numShufflePartitions
+        // the conversion to catalyst expressions above produces SortOrder expressions
+        // for OrderedDistribution and generic expressions for ClusteredDistribution
+        // this allows RepartitionByExpression to pick either range or hash partitioning
+        RepartitionByExpression(distribution, query, numShufflePartitions)
```

Comment:
Is it possible the required distribution be changed later by other optimization? The distribution requirement from data source is a hard requirement? Once if the distribution is changed and not matched the requirement, how will data source react to it?

Reply:
We are inserting repartition/sort nodes directly before writing so my assumption that Spark will only remove them if the incoming plan already satisfies these requirements. `WriteDistributionAndOrderingSuite` is kind of meant for testing that. Do you have ideas when this assumption will break, @viirya?

Reply:
Oh, I see. Mis-reading this part. Looks fine. Thanks.